updater/lib/dependabot/dependency_snapshot.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+require "base64"
+require "dependabot/file_parsers"
+
+# This class describes the dependencies obtained from a project at a specific commit SHA
+# including both the Dependabot::DependencyFile objects at that reference as well as
+# means to parse them into a set of Dependabot::Dependency objects.
+#
+# This class is the input for a Dependabot::Updater process with Dependabot::DependencyChange
+# representing the output.
+module Dependabot
+  class DependencySnapshot
+    # TODO: Enforce non-nil values for job_definition["base64_dependency_files"]
+    #       and job_definition["base_commit_sha"]
+    #
+    # We historically tolerate nil values for both these keys from the `job_definition`
+    # but it doesn't seem like we should. Rather than introduce a behaviour change
+    # as part of the change introducing this class, let's do it as a follow-up.
+    def self.create_from_job_definition(job:, job_definition:)
+      decoded_dependency_files = job_definition.fetch("base64_dependency_files", []).map do |a|
+        file = Dependabot::DependencyFile.new(**a.transform_keys(&:to_sym))
+        file.content = Base64.decode64(file.content).force_encoding("utf-8") unless file.binary? && !file.deleted?
+        file
+      end
+
+      new(
+        job: job,
+        base_commit_sha: job_definition.fetch("base_commit_sha", nil),
+        dependency_files: decoded_dependency_files
+      )
+    end
+
+    attr_reader :base_commit_sha, :dependency_files
+
+    def initialize(job:, base_commit_sha:, dependency_files:)
+      @job = job
+      @base_commit_sha = base_commit_sha
+      @dependency_files = dependency_files
+    end
+
+    def dependencies
+      return @dependencies if defined?(@dependencies)
+
+      parse_files!
+    end
+
+    private
+
+    attr_reader :job
+
+    # TODO: Parse files during instantiation?
+    #
+    # To avoid having to re-home Dependabot::Updater#handle_parser_error,
+    # we perform the parsing lazily when the `dependencies` method is first
+    # referenced.
+    #
+    # We have some unusual behaviour where we handle a parse error by
+    # returning an empty dependency array in Dependabot::Updater#dependencies
+    # in order to 'fall through' to an error outcome elsewhere in the class.
+    #
+    # Given this uncertainity, and the need to significantly refactor tests,
+    # it makes sense to introduce this shim and then deal with the call
+    # site once we've split out the downstream behaviour in the updater.
+    #
+    def parse_files!
+      @dependencies = dependency_file_parser.parse
+    end
```

Comment:
🤔 Would it make sense to expose a method that signals that parsing has completed? I understand defining `@dependencies` signals this internally, and that we may not want to have collaborators checking this externally. Maybe something like a result object would be appropriate, given your note about handling parse errors.

None of this is necessary, but _maybe_ it would be useful? I'm interested in your thoughts as we're still paving the path for further changes.

Reply:
> we may not want to have collaborators checking this externally.

I think the call order we have right now means we _can_ proceed if parsing fails, but from reasoning through the code I don't think we ever actually should? i.e. there shouldn't be any collaboration if parsing fails, it's always a sign to stop the job.

https://github.com/dependabot/dependabot-core/blob/8f9e4734020e0029a9ec53e1d38163bbcc6235c7/updater/lib/dependabot/updater.rb#L71-L77

In the existing behaviour, if parsing fails, we get an empty value for "dependencies" at the code above.

For a new update, `[]` means we don't iterate and the job silently ends.

For updating an existing PR, we end up here:
https://github.com/dependabot/dependabot-core/blob/8f9e4734020e0029a9ec53e1d38163bbcc6235c7/updater/lib/dependabot/updater.rb#L134-L146

This will nearly always result in `[] != [$at_least_one_dep]`, so we exit closing the existing PR.

This is feels an awful lot like cleanly closing the job accident/on a weak signal. We do always place a parser error on the UpdateJob, so the user should at least see the job failed due to unparsable files, but I think it also means we are likely closing "rebased" PRs on parse errors and that might be a bug.